### PR TITLE
Monomorphize functions passed as arguments

### LINKED ISSUE #544: when monomorphizing, also monomorphize global function/value bindings
e.g. consider

```
(coalton-codegen
  (declare square (Num :t => :t -> :t))
  (define (square x) (* x x))

  (declare norm^2 ((Num :t) => (List :t) -> :t))
  (define (norm^2 x)
    (reduce + 0 (map square x)))

  (monomorphize)
  (declare norm^2-df ((List Double-Float) -> Double-Float))
  (define norm^2-df norm^2))
```

In this case, a full polymorphic call to `+` and `square` will still occur.

This would help make monomorphized functions faster.

### PR DESCRIPTION
**Pull request by eliaslfox**
Fixes #544

### CODE REVIEW

**Comment by eliaslfox** [comment]
```
(coalton-codegen
  (declare square (Num :t => :t -> :t))
  (define (square x) (* x x))

  (monomorphize)
  (declare norm^2 (List Double-Float -> Double-Float))
  (define (norm^2 x)
    (reduce + 0 (map square x))))

(COMMON-LISP:DEFUN NORM^2 (X-2187)
   (COMMON-LISP:DECLARE (COMMON-LISP:IGNORABLE X-2187))
   (|REDUCE#2189| ARITH::|INSTANCE/NUM DOUBLE-FLOAT-+|
    (ARITH::|INSTANCE/NUM DOUBLE-FLOAT-FROMINT| 0)
    (LIST::|INSTANCE/FUNCTOR LIST-MAP| |SQUARE#2188| X-2187)))
```

File: src/codegen/monomorphize.lisp [hunk 1]
```diff
@@ -182,8 +183,14 @@ recompilation, and also maintains a stack of uncompiled candidates."
 
     ;; Fully propigating arguments could change the ordering of side effects
     ;; Exit early to avoid this
+    ;;
+    ;; Unless the function is known to have more arguments than are being applied
     (when (every #'propagatable-p new-args)
-      (return-from valid-candidate-p))
+      (let ((function (tc:lookup-function env name :no-error t)))
+        (unless function
+          (return-from valid-candidate-p))
```
Comment (stylewarning) [inline src/codegen/monomorphize.lisp:191]: can you explicitly return NIL here and in the other cases ?

**Review by stylewarning — APPROVED** [review]
one small req

File: src/codegen/monomorphize.lisp (hunk 1 above)
Comment (eliaslfox) [reply src/codegen/monomorphize.lisp:191]: Is that standard style in common lisp? I feel like it just add line noise

Comment (stylewarning) [reply src/codegen/monomorphize.lisp:191]: I think it's like `when`/`unless` vs `cond`/`if`. If the return value matters (in this case it does), them being explicit is better than implicit. If the return value doesn't matter (a function just for effect), then eliding the return value is OK.

A non-expert Lisp programmer might also return `(return-from f)` as `f` returning `(values)`, which is not correct, but not an unreasonable assumption.

**Pull request merged**